### ISSUE TITLE
Solaris 11: KDC generates unusable principals

### ISSUE THREAD

**Issue by quanah** [heimdal/heimdal#441]
When testing out Heimdal on Solaris 11, we have found that Heimdal is unable to create/extract principals in the KDC.  This works w/o issue on all our Linux platforms.

```
/opt/symas/bin/kadmin -l init --realm-max-ticket-life=unlimited --realm-max-renewable-life=unlimited TRUPP.SYMAS.NET
kadmin: rc4 8: EVP_CipherInit_ex encrypt

/opt/symas/bin/kadmin -l add --password=bob --use-defaults bob@TRUPP.SYMAS.NET
kadmin: rc4 8: EVP_CipherInit_ex encrypt

/opt/symas/bin/kadmin -l add --random-key --use-defaults ldap/test.trupp.symas.net@TRUPP.SYMAS.NET
kadmin: rc4 8: EVP_CipherInit_ex encrypt

/opt/symas/bin/kadmin -l ext ldap/test.trupp.symas.net@TRUPP.SYMAS.NET
kadmin: ext ldap/test.trupp.symas.net@TRUPP.SYMAS.NET: Principal does not exist

/opt/symas/bin/kadmin -l ext bob@TRUPP.SYMAS.NET
kadmin: ext bob@TRUPP.SYMAS.NET: Principal does not exist

/opt/symas/bin/kinit -k bob@TRUPP.SYMAS.NET
kinit: krb5_init_creds_set_keytab: Failed to find bob@TRUPP.SYMAS.NET in keytab FILE:/opt/symas/etc/heimdal/krb5.keytab (unknown enctype)

```

**Comment by quanah** [comment]
All tests except the following pass in the test suite:

```
FAIL: check-iprop
```
in tests/kdc

test-suite.log shows:

```
FAIL: check-iprop
=================

Creating database
Test log recovery
Starting kdc
starting master
starting slave
Waiting for ipropd-slave to start, looking logfile messages.log
Have waited 2 seconds
Have waited 4 seconds
Have waited 6 seconds
Have waited 8 seconds
Have waited 10 seconds
Have waited 12 seconds
Have waited 14 seconds
Have waited 16 seconds
Have waited 18 seconds
Have waited 20 seconds
Have waited 22 seconds
Have waited 24 seconds
Have waited 26 seconds
Have waited 28 seconds
Have waited 30 seconds
Have waited 32 seconds
Have waited 34 seconds
Have waited 36 seconds
Have waited 38 seconds
Have waited 40 seconds
Have waited 42 seconds
Have waited 44 seconds
Have waited 46 seconds
Have waited 48 seconds
Have waited 50 seconds
Have waited 52 seconds
Have waited 54 seconds
Have waited 56 seconds
Have waited 58 seconds
Have waited 60 seconds
Have waited 62 seconds
Have waited 64 seconds
Have waited 66 seconds
Waited for 65 for the ipropd-slave to start, and it didnt happen
killing ipropd s + m + kdc
2018-11-20T07:29:30 error message: Matching credential (krbtgt/TEST.H5L.SE@TEST.H5L.SE) not found: -1765328243
2018-11-20T07:29:30 error message: Matching credential (krbtgt/RB.SYMAS.NET@TEST.H5L.SE) not found: -1765328243
2018-11-20T07:29:30 error message: Matching credential (krbtgt/TEST.H5L.SE@TEST.H5L.SE) not found: -1765328243
2018-11-20T07:29:30 error message: Matching credential (iprop/s11build.rb.symas.net@RB.SYMAS.NET) not found: -1765328243
2018-11-20T07:29:30 krb5_sendauth: Matching credential (iprop/s11build.rb.symas.net@RB.SYMAS.NET) not found
2018-11-20T07:29:30 disconnected for server
2018-11-20T07:29:30 sleeping 4 seconds before retrying to connect
2018-11-20T07:29:30 krb5_recvauth: End of file
2018-11-20T07:29:31 ipropd-master terminated
2018-11-20T07:29:31 ipropd-slave terminated
master down at 2018-11-20T07:29:31
FAIL check-iprop (exit status: 1)
```

**Comment by quanah** [comment]
@lhoward are there any additional details I can provide?

**Comment by lhoward** [comment]
My bad, been swamped. I’ll get to it soon.

**Comment by lhoward** [comment]
check-iprop passes for me on Solaris 11.3. Are you running master (specifically do you have 934d5e09 applied)?

**Comment by lhoward** [comment]
The kadmin examples in your initial report work for me, using master on Solaris 11.3. So... let's just check first you are on master?

```
% uname -a
SunOS storgata.lukktone.com 5.11 11.3 i86pc i386 i86pc
```

**Comment by quanah** [comment]
Yep:

```
quanah@sol11-build:~$ uname -a
SunOS sol11-build.quanah.org 5.11 11.3 i86pc i386 i86pc Solaris
```

Which compiler are you using?  I'm using:

```
gcc (GCC) 4.8.2
```

**Comment by quanah** [comment]
Oh, if you mean Heimdal master, no.  I need a working Heimdal release.  So at the moment, I'm on Heimdal 7.5 with backported patches for Solaris, since a number of the fixes to get Heimdal to even compile on Solaris are missing from the current release.

**Comment by quanah** [comment]
@lhoward 

In Heimdal master, I hit the following test failure on Solaris 11.3 after which the test suite stops:

```
==============================================
   Heimdal 7.99.1: lib/roken/test-suite.log
==============================================

# TOTAL: 14
# PASS:  13
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test-auxval
=================

test-auxval: No ELF auxv types to inject
test-auxval: rk_injectprocauxv() failed
auxv type 2008 -> 4278185866
auxv type 2014 -> 4278185872
auxv type 3 -> 134545460
auxv type 4 -> 32
auxv type 5 -> 5
auxv type 9 -> 134550304
auxv type 2016 -> 4269776896
auxv type 7 -> 4269477888
auxv type 8 -> 0
auxv type 6 -> 4096
auxv type 2017 -> 258
auxv type 2009 -> 3365999
auxv type 2023 -> 0
FAIL test-auxval (exit status: 1)
```

**Comment by lhoward** [comment]
Have you back-ported the recent changes to lib/hcrypto/evp-pkcs11.c?

**Comment by lhoward** [comment]
> Which compiler are you using? I'm using:
> 
> gcc (GCC) 4.8.2
I’m using the Sun compiler.

**Comment by quanah** [comment]
@lhoward 

The test failure with test-auxval is with master as of today at 21:08 Zulu.  There shouldn't be anything to backport?

**Comment by lhoward** [comment]
I’ll check when I get in. Either something changed in master since yesterday, or it’s a compiler (unlikely?) or system difference.

**Comment by quanah** [comment]
@lhoward This same test also fails for me on ubuntu 18 64-bit, using gcc:

```
gcc --version
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

It logs:

```
test-auxval: rk_getauxval(16) should have been 126548989, was 2
auxv type 33 -> 140732995850240
FAIL test-auxval (exit status: 1)
```

**Comment by lhoward** [comment]
OK, that test is also failing for me, I had only run the lib/hcrypto, lib/krb5 and tests/kdc tests so that would explain that one. I shall investigate. Were the other tests passing for you?

**Comment by quanah** [comment]
@lhoward I'll start poking at other portions of the test suite and see if they also have failures.

**Comment by lhoward** [comment]
It all succeeded for me now :)

**Comment by quanah** [comment]
@lhoward hm, I can't see how that would fix the issue with Ubuntu Linux though. ;)

**Comment by lhoward** [comment]
I'll commit the fix above once @nicowilliams approves it.

**Comment by quanah** [comment]
After applying the suggested patch, not only did Heimdal build to completion and all tests pass, but my test using SASL/GSSAPI with OpenLDAP also worked for the first time ever on Solaris 11.

**Comment by nicowilliams** [comment]
@lhoward That should work, however, so would making sure that `HAVE_ISSETUGID` gets set by the autoconf system and that we use it.  What's actually happening with that??

**Comment by lhoward** [comment]
According to your notes in issuid.c, issetugid() cannot be trusted on all platforms. So it depends on whether you want to put platform specific tests in autoconf or the code for issetugid() that meets your criteria.

**Comment by nicowilliams** [comment]
@lhoward It absolutely can be trusted on these systems: Solaris, Illumos, and OpenBSD, (but not on any others)

**Comment by nicowilliams** [comment]
```
     * On OpenBSD (where issetugid() originated), Illumos derivatives,
     * and Solaris, issetugid() returns true IFF the program exec()ed
     * was set-uid or set-gid.
```

I've actually talked to NetBSD developers about adopting the Solaris/Illumos/OpenBSD semantics, as the NetBSD/FreeBSD/DragonFly semantics are useless.  But they seem to believe it'd be a backwards-incompatible change :(

**Comment by nicowilliams** [comment]
Anyways, your diff would do, yes.

**Comment by lhoward** [comment]
Fixed in ec88576a

**Comment by nicowilliams** [comment]
@lhoward What about the issue in the original description?

```
/opt/symas/bin/kadmin -l init --realm-max-ticket-life=unlimited --realm-max-renewable-life=unlimited TRUPP.SYMAS.NET
kadmin: rc4 8: EVP_CipherInit_ex encrypt
```

**Comment by quanah** [comment]
@nicowilliams Sorry, I was covering that in this comment:

"my test using SASL/GSSAPI with OpenLDAP also worked for the first time ever on Solaris 11."

I.e., part of that test is creating example principals like bob@<TESTREALM> and then binding to LDAP via SASL/GSSAPI.

**Comment by quanah** [comment]
@nicowilliams The report was against Heimdal 7.5.0.  master has had a lot of fixes in this area since, apparently.  The issuid() stuff fixed the problems I hit when building master against Solaris 11.

**Comment by nicowilliams** [comment]
@quanah Ah, OK, thanks!

**Comment by lhoward** [comment]
issuid() just caused a test suite failure, no operational issues I noticed

**Comment by lhoward** [comment]
The original issue was I suspect fixed by my PKCS#11 changes from earlier this year.
